### PR TITLE
Bug 1882405 - Page Settings Data Structure Should be `Val`

### PR DESCRIPTION
**Pull request by ohall-m**
* A subtle bug was introduced when page settings were moved from `val` to `var` for updating page settings.
* Everything works as expected with the reducer updating etc., however changes on page settings are not observed in compose.

### Pull Request checklist

- [X] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [X] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1882405

### CODE REVIEW

**Comment by ohall-m** [comment]
Had trouble finding a way to write an automated test, closest thing was [these examples](https://searchfox.org/mozilla-mobile/search?q=store.observe&path=Test.kt&case=false&regexp=false) I found. I have a sense this bug was more a logic error misunderstanding on my part initially too. The [actions do have appropriate tests](https://searchfox.org/mozilla-mobile/search?q=UpdatePageSettingAction&path=*TranslationsActionTest.kt&case=false&regexp=false).

Did manually test against patch #5526:

Without Patch:

https://github.com/mozilla-mobile/firefox-android/assets/93344309/78ee2e00-6e32-4bea-9892-573e65b2b091

With Patch:

https://github.com/mozilla-mobile/firefox-android/assets/93344309/87c1ec67-f5ee-4dd2-aabc-8f9ad88b85d0

File: android-components/components/browser/state/src/test/java/mozilla/components/browser/state/action/TranslationsActionTest.kt [hunk 1]
```diff
@@ -690,6 +690,7 @@ class TranslationsActionTest {
         assertTrue(tabState().translationsState.pageSettings?.neverTranslateSite!!)
     }
 
+    @Test
```
Comment (ohall-m) [inline android-components/components/browser/state/src/test/java/mozilla/components/browser/state/action/TranslationsActionTest.kt:693]: Just happened to see it was missing the annotation when checking tests.

**Review by gabrielluong — APPROVED** [review]

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt [hunk 2]
```diff
@@ -263,51 +263,67 @@ internal object TranslationsStateReducer {
             }
 
         is TranslationsAction.UpdatePageSettingAction -> {
-            var pageSettings = state.findTab(action.tabId)?.translationsState?.pageSettings
+            var currentPageSettings = state.findTab(action.tabId)?.translationsState?.pageSettings
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt]: Thanks, much more kotlin-y!

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt [hunk 3]
```diff
@@ -263,51 +263,67 @@ internal object TranslationsStateReducer {
...
                         // Always and never translate sites are always opposites when the other is true.
-                        pageSettings.alwaysTranslateLanguage = false
+                        alwaysTranslateLang = false
                     }
 
+                    val newSettings =
+                        currentPageSettings.copy(
+                            alwaysTranslateLanguage = alwaysTranslateLang,
+                            neverTranslateLanguage = neverTranslateLang,
+                        )
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt]: Inlined!

File: android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt [hunk 4]
```diff
@@ -263,51 +263,67 @@ internal object TranslationsStateReducer {
...
                         )
                     }
                 }
+
                 TranslationPageSettingOperation.UPDATE_NEVER_TRANSLATE_LANGUAGE -> {
-                    pageSettings.neverTranslateLanguage = action.setting
-                    if (action.setting) {
+                    var alwaysTranslateLang = currentPageSettings.alwaysTranslateLanguage
+                    val neverTranslateLang = action.setting
+                    if (neverTranslateLang) {
```
[the remark this answers is not in the document]
Comment (ohall-m) [reply android-components/components/browser/state/src/main/java/mozilla/components/browser/state/reducer/TranslationsStateReducer.kt]: Updated!

**Pull request merged**